### PR TITLE
feat: custom neutron governance

### PR DESCRIPTION
**Pull request by quasisamurai**
**Task** 
-  [link](https://p2pvalidator.atlassian.net/browse/NTRN-139)

 **Motivation**
 - in our network there will be no classic cosmos validators, as well as staking will be turned off. So we still need some governance, preferably w default cosmos interface. And this pr(s) is/are a solution.
 
 
**What's happening here?** 
-  Because instead of staking, we plan to use a smart contract that freezes the user's funds, giving him the opportunity to take part in the governance, here I introduce a patched cosmos  `gov`  module which has 3 main features 

1.  Custom `tally` logic, which is calculates voting power based on smart contract ( see [PR](https://github.com/neutron-org/neutron-dao/pull/1) )
2.  Since it requires a smart contract as part of our blockchain "out of the box", `cmd` and network's `init.sh` has been patched to include our contract in `genesis.json`
3.   `wasmkeeper` also added to renewed `gov` module due the `tally` logic requires smart contract interaction

issues:
- where we should store contract wasm binary? mb `contracts` dir?

this PR is strongly related to [this](https://github.com/neutron-org/neutron-dao/pull/1)

### CODE REVIEW

File: network/init.sh [hunk 1]
```diff
@@ -69,6 +71,15 @@ $BINARY gentx val2 7000000000stake --home $CHAIN_DIR/$CHAINID_2 --chain-id $CHAI
 $BINARY collect-gentxs --home $CHAIN_DIR/$CHAINID_1
 $BINARY collect-gentxs --home $CHAIN_DIR/$CHAINID_2
 
+#TODO: dynamic addresses
+echo "Initializing dao contract in genesis..."
+# Upload the queries contract
```
Comment (zavgorodnii) [inline network/init.sh]: something is wrong with these comments (copy pasted?)

File: network/init.sh [hunk 2]
```diff
@@ -69,6 +71,15 @@ $BINARY gentx val2 7000000000stake --home $CHAIN_DIR/$CHAINID_2 --chain-id $CHAI
...
+$BINARY add-wasm-message store ${DAO_CONTRACT} --output json  --run-as neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u --home $CHAIN_DIR/$CHAINID_1
+# Instantiate the contract
+INIT_CONTRACT='{"owner":"neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u"}'
```
Comment (zavgorodnii) [inline network/init.sh]: move this address to a variable please

File: x/gov/abci.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+		passes, burnDeposits, tallyResults := keeper.Tally(ctx, proposal) // our custom implementation of tally logics
+
+		if burnDeposits {
+			keeper.DeleteDeposits(ctx, proposal.ProposalId)
+		} else {
+			keeper.RefundDeposits(ctx, proposal.ProposalId)
+		}
+
+		if passes {
```
Comment (zavgorodnii) [inline x/gov/abci.go:62]: (note to self) this is a 0.45 version of governance, we'll need to patch this module to 0.46 if we choose to switch sdks

**Comment by zavgorodnii** [comment]
please create a temporary dir `./contracts` and move the dao contract binary there

File: network/init.sh (hunk 2 above)
Comment (quasisamurai) [reply network/init.sh]: yeah, todo is about this

File: network/init.sh [hunk 4]
```diff
@@ -4,6 +4,8 @@ BINARY=neutrond
 CHAIN_DIR=./data
 CHAINID_1=test-1
 CHAINID_2=test-2
+ADDRESS=neutron1mjk79fjjgpplak5wq838w0yd982gzkyf8fxu8u
```
Comment (zavgorodnii) [inline network/init.sh]: what address? please make this more explicit

**Comment by zavgorodnii** [comment]
looks okay, tested with the script from https://github.com/neutron-org/neutron-dao/pull/1, the proposal passes

**Review by zavgorodnii — DISMISSED** [review]

**Review by pr0n00gler — DISMISSED** [review]

File: x/gov/keeper/tally.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	wasmkeeper "github.com/CosmWasm/wasmd/x/wasm/keeper"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+)
+
+// DefaultContractAddr is the wasm contract address generated by code ID 1 and instance ID 1.
+//
+// In other words, the first ever contract to be deployed on this chain will necessarily have this address.
+var DefaultContractAddr = wasmkeeper.BuildContractAddress(1, 1)
```
Comment (swelf19) [inline x/gov/keeper/tally.go:13]: do we want to require that wasmdao must be baked into genesis, if so, is it possible to check it on the node start

**Review by swelf19 — DISMISSED** [review]

File: x/gov/README.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+The original gov module computes voting power on proposals using each user's delegations. Since the Neutron's plan is to not use standard staking & validators, it's necessary to remove rudimentary usage of staking module. Instead of this, however, we are introduced an alternative
+## Voting power is now calculated and stored in smart-contract
+
+We use cosm-wasm contract which implements several methods:
+```rust
+pub fn query_voting_power(deps: Deps, user_addr: Addr) -> StdResult<VotingPowerResponse> {...}
+```
+
+```rust
+pub fn query_voting_powers(deps: Deps) -> StdResult<Vec<VotingPowerResponse>>  {...}
```
Comment (swelf19) [inline x/gov/README.md:22]: i think а pagination should be implemented for the method. But we can implement it in a separate task.

**Review by swelf19 — APPROVED** [review]

File: x/gov/README.md (hunk 6 above)
Comment (quasisamurai) [reply x/gov/README.md:22]: agreed

**Pull request merged**